src/main/scala/io/iohk/ethereum/consensus/ConsensusImpl.scala

```diff
@@ -28,226 +20,168 @@ import io.iohk.ethereum.ledger.BlockData
           case None         => returnNoTotalDifficulty(bestBlock)
         }
       case None => returnNoBestBlock()
     }
-  }
 
-  private def handleBlockImport(block: Block, bestBlock: Block, weight: ChainWeight)(implicit
+  private def handleBranchImport(
+      branch: NonEmptyList[Block],
+      currentBestBlock: Block,
+      currentBestBlockWeight: ChainWeight
+  )(implicit
       blockExecutionScheduler: Scheduler,
       blockchainConfig: BlockchainConfig
-  ): Task[BlockImportResult] = {
-    val importResult = if (isPossibleNewBestBlock(block.header, bestBlock.header)) {
-      importToTop(block, bestBlock, weight)
-    } else {
-      reorganiseOrEnqueue(block, bestBlock, weight)
+  ): Task[ConsensusResult] = {
+    val parentHash = branch.head.header.parentHash
+    val parentWeight = blockchain
+      .getChainWeightByHash(parentHash)
+      .getOrElse(throw new Error("Inconsistent database"))
+
+    val consensusResult: Task[ConsensusResult] =
+      if (newBranchWeight(branch, parentWeight) > currentBestBlockWeight) {
+        if (currentBestBlock.isParentOf(branch.head)) {
+          Task.evalOnce(importToTop(branch, currentBestBlockWeight)).executeOn(blockExecutionScheduler)
+        } else {
+          Task.evalOnce(reorganise(branch, parentWeight, parentHash)).executeOn(blockExecutionScheduler)
+        }
+      } else {
+        Task.now(KeptCurrentBestBranch)
+      }
+
+    consensusResult.foreach(measureBlockMetrics)
+    consensusResult
+  }
+
+  private def importToTop(branch: NonEmptyList[Block], currentBestBlockWeight: ChainWeight)(implicit
+      blockchainConfig: BlockchainConfig
+  ): ConsensusResult =
+    blockExecution.executeAndValidateBlocks(branch.toList, currentBestBlockWeight) match {
+      case (importedBlocks, None) =>
+        importedBlocks.lastOption.foreach(b => blockchain.saveBestKnownBlocks(b.block.hash, b.block.number))
+
+        ExtendedCurrentBestBranch(importedBlocks)
+      case (_, Some(MPTError(reason))) if reason.isInstanceOf[MissingNodeException] =>
+        ConsensusErrorDueToMissingNode(Nil, reason.asInstanceOf[MissingNodeException])
+      case (Nil, Some(error)) =>
+        BranchExecutionFailure(Nil, branch.head.header.hash, error.toString)
+      case (importedBlocks, Some(error)) =>
+        importedBlocks.lastOption.foreach(b => blockchain.saveBestKnownBlocks(b.block.hash, b.block.number))
+        val failingBlock = branch.toList.drop(importedBlocks.length).head
+        ExtendedCurrentBestBranchPartially(
+          importedBlocks,
+          BranchExecutionFailure(Nil, failingBlock.hash, error.toString)
+        )
     }
-    importResult.foreach(measureBlockMetrics)
-    importResult
+
+  private def reorganise(newBranch: NonEmptyList[Block], parentWeight: ChainWeight, parentHash: ByteString)(implicit
+      blockchainConfig: BlockchainConfig
+  ): ConsensusResult = {
+
+    val bestNumber = blockchainReader.getBestBlockNumber()
```

Comment:
The caller method has the info on the best block, so you can just pass it as an argument here